### ISSUE TITLE
Uncaught exception handler loses exception object, stack trace, message

### ISSUE THREAD

**Issue by niloc132** [gwtproject/gwt#9734]
**GWT version: 2.8.2, 2.9.0, HEAD
**Browser (with version):** latest Chrome
**Operating System:** Tested on linux, mac, windows

---

##### Description

At some point, Chrome stopped relaying the details of an `onerror` event when the error happens in a script file loaded cross-origin. This does not apply to old-style `$entry`-wrapped stack traces, but only to errors caught by the uncaught exception handler from the gwt iframe onerror handler.

When this occurs, the onerror handler in Impl.java should get the parameters "msg, url, line, column, error", but instead receives "Script error.", "", 0, 0, null for the parameters, so the actual exception cannot be extracted.

Note that this also applies to the "unhandledrejection" event, which should contain the rejected details, but instead doesn't go off at all.

According to https://developer.mozilla.org/en-US/docs/Web/HTML/Element/script#attr-crossorigin, an attribute can be set on the script tag when it is created to allow the onerror handler to receive the details of the exception when it happens. The server must opt-in as well, as documented in that link, by setting the appropriate CORS headers. Note that the CodeServer class `com.google.gwt.dev.codeserver.WebServer` already correctly sets this header for served JS content (I think at the time this was required for Firefox, so the header was added, but since then Chrome additionally required the attribute on the script tag?):
```
    // Wrap the response to send the extra headers.
    return new Response() {
      @Override
      public void send(HttpServletRequest request, HttpServletResponse response, TreeLogger logger)
          throws IOException {
        // TODO: why do we need this? Looks like Ray added it a long time ago.
        response.setHeader("Access-Control-Allow-Origin", "*");
```

##### Steps to reproduce
The simple case:
* Attach a simple error handler, just GWT.setUncaughtExceptionHandler is sufficient, and log the exception type or message
* In a non-JSNI callback, throw an exception of some kind. If adding an onerror handler manually, be sure to add it to the GWT iframe, or it won't go off at all.
* Load the app from one ~domain~ origin with SuperDevMode, so that the compiled JS is served from a different ~domain~ origin, observe the logged message

Rejected promise:
* Add a "unhandledrejection" event handler to the gwt iframe
* Reject any promise, probably in an event handler of some kind to make it easily reproducible
* Load the app as above - do not set breakpoints in the rejection, sometimes Chrome will end up actually calling the event handler if this happens.

See https://github.com/vertispan/unhandled-promise-rejection for a runnable sample and workaround branch for both exceptions and promises. Note that the workaround only addresses the simple case, the submitted patch also modifies every other injected gwt script tag I could find, so that we are sure that all resources can be loaded cross origin (IF headers are set by the server) and have exceptions function properly.

##### Known workarounds
For the standard `xsiframe` linker, you can patch the `com/google/gwt/core/ext/linker/impl/installScriptDirect.js` file in your own project to set the `crossOrigin` property of the created script tag. The file is only 34 lines long before this change, so here's the complete file with the added line and a comment:
```
// Installs the script directly, by simply appending a script tag with the
// src set to the correct location to the install location.
function installScript(filename) {
  // Provides the setupWaitForBodyLoad()function
  __WAIT_FOR_BODY_LOADED__
  
  function installCode(code) {
    var doc = getInstallLocationDoc();
    var docbody = doc.body;
    var script = doc.createElement('script');
    script.language='javascript';
    script.crossOrigin='';// THIS IS THE ADDED LINE TO PREVENT THIS BUG
    script.src = code;
    if (__MODULE_FUNC__.__errFn) {
      script.onerror = function() {
        __MODULE_FUNC__.__errFn('__MODULE_FUNC__', new Error("Failed to load " + code));
      }
    }
    docbody.appendChild(script);
    sendStats('moduleStartup', 'scriptTagAdded');
  }

  // Start measuring from the time the caller asked for this file,
  // for consistency with installScriptEarlyDownload.js.
  // The elapsed time will include waiting for the body.
  sendStats('moduleStartup', 'moduleRequested');

  // Just pass along the filename so that a script tag can be installed in the
  // iframe to download it.  Since we will be adding the iframe to the body,
  // we still need to wait for the body to load before going forward.
  setupWaitForBodyLoad(function() {
    installCode(filename);
  });
}
```

##### Links to further discussions
https://gitter.im/DominoKit/domino?at=60ef0045049a0c3e5372b98a

**Comment by niloc132** [comment]
Proposed patch at https://gwt-review.googlesource.com/c/gwt/+/23500, applying this attribute not only in the file mentioned above, but any JS used to create a script tag that could be inside an iframe.

**Comment by zbynek** [comment]
@niloc132 with this it's no longer possible to open pages over `file://` protocol with the JS also loading over `file://`.

```
Access to script at 'file:///dir/foo.cache.js' from origin 'null' has been blocked by CORS policy: 
Cross origin requests are only supported for protocol schemes: http, data, isolated-app, chrome-extension, chrome, https, chrome-untrusted.
```

**Comment by niloc132** [comment]
(Should have been closed as part of 2.10, closing now)
